### ISSUE TITLE
Could I submit pull request for enabling Allow/Deny validation for Restivus?

### ISSUE THREAD

**Issue by VansonLeung** [kahmali/meteor-restivus#163]
I have just made myself Allow/Deny Validation for Restivus, because separating methods & allow/deny validation feels stupid for me, and feels more stupid when it is also true for Restivus. That means if I want to use Allow/Deny for CMS side, it implies I have to reinvent the wheel for Mobile Apps API side.

Solved my own problem by the codes below. Tested OK compatible with 

aldeed's Simple Schema autoValue , 
Collection2 Allow & Deny, and 
ongoworks:meteor-security.

Could I submit a pull request??

Codes below:
```
SERVER_GENERIC_VALIDATE = function(type, userId, doc, fields, modifiers, collection, collectionName, isDebug) {
  // if collection == null, then try collectionName
  try {
    if (!collection) collection = global[collectionName];
    if (!collection) {
      throw "Error, collection / collection name not found";
    }

    // if type == update, requires collection & doc._id to exist
    if (type === 'update') {
      if (doc._id) {
        doc = collection.findOne(doc._id);
      } else {
        throw "Error, update: doc._id not specified";
      }
    }


    // bootstrap autovalues
    this.isInsert = (type === "insert");
    this.isUpdate = (type === "update");

    for (var autoValue in collection._autoValues)
    {
      var value = collection._autoValues[autoValue].bind(this)();
      if (value != undefined)
      {
        doc[autoValue] = value;
      }
    }

    if (isDebug)
    {
      console.log("Collection: " + collection._name);
      console.log("Type: " + type);
      console.log("UserId: " + userId);
      console.log("Doc: " + doc);
      console.log("Modifiers: " + modifiers);
    }

    try {
      var allows = collection._validators[type].allow;
      var denies = collection._validators[type].deny;
      if (allows.length <= 0 && denies.length <= 0)
      {
        return {statusCode: 403, body: {status: 'fail', message: 'Access Denied'}}
      }
      for (var i = 0; i < allows.length; i++) {
        var result = allows[i](
            userId,
            doc,
            fields,
            modifiers
        );
        if (!result) {
          return {statusCode: 403, body: {status: 'fail', message: 'Access Denied'}}
        }
      }
      for (var i = 0; i < denies.length; i++)
      {
        var result = denies[i](
            userId,
            doc,
            fields,
            modifiers
        );
        if (result) {
          return {statusCode: 403, body: {status: 'fail', message: 'Access Denied'}}
        }
      }
    } catch (e) {
      if (isDebug) console.log(e);
      return {statusCode: 401, body: {status: 'fail', message: e}}
    }

  } catch (e) {
    if (isDebug) console.log(e);
    return {statusCode: 400, body: {status: 'fail', message: 'Internal Server Error'}}
  }
};


SERVER_GENERIC_VALIDATE_UNITTEST = function(expected_error, type, userId, doc, fields, modifiers, collection, collectionName, isDebug) {
  var result = SERVER_GENERIC_VALIDATE.call(this, type, userId, doc, fields, modifiers, collection, collectionName, isDebug);

  if (expected_error) {
    if (result) console.log(""+Roles.getRolesForUser(userId)+" "+type+" "+collection._name+" - TEST OK");
    else console.log(""+Roles.getRolesForUser(userId)+" "+type+" "+collection._name+" - TEST FAIL");
  } else {
    if (!result) console.log(""+Roles.getRolesForUser(userId)+" "+type+" "+collection._name+" - TEST OK");
    else console.log(""+Roles.getRolesForUser(userId)+" "+type+" "+collection._name+" - TEST FAIL");
  }
}
```

**Comment by kahmali** [comment]
Hi @VansonLeung,

I'm no security expert, so I'll just have to take MDGs word that trying to validate updates via allow/deny is a pretty big security risk, so this isn't something that I would pull into the core of Restivus. From the [Meteor Guide's section on security](http://guide.meteor.com/security.html):

>In this guide, we’re going to take a strong position that using allow or deny to run MongoDB queries directly from the client is not a good idea. The main reason is that it is hard to follow the principles outlined above. It’s extremely difficult to validate the complete space of possible MongoDB operators, which could potentially grow over time with new versions of MongoDB.

However, if folks were willing to accept the risk, then I'd love to provide a way for them to use this as an add-on package. My goal is to make Restivus a lot more modular. Have you tried adding this functionality from a separate package? If you need me to expose an API for you to add this on then I can certainly do that. Or maybe _you_ can in another PR, since the time I have for Restivus is extremely limited (read: nonexistent). If you give it a shot, let me know what type of access you need to the Restivus internals, and I'll work on exposing an API for you as soon as possible. If you need Restivus to expose an API, just run any thoughts by me and we'll figure something out. I know for sure it's possible, I just want to do it cleanly from the start. 

Let me know if you plan to follow up with this. If not, we can close the issue.

**Comment by VansonLeung** [comment]
I have implemented my security middleware by calling the code above,

and my allow/deny rule references is set in Server side too:


```js
doACL = function() {
  if (Meteor.isServer) {

    Security.defineMethod('isAdmin', {
      fetch: [],
      transform: null,
      deny: function (type, arg, userId, doc) {
        var deny = !Roles.userIsInRole(userId, ['superadmin', 'admin']);
        return deny;
      }
    });


    Security.defineMethod('isClient', {
      fetch: [],
      transform: null,
      deny: function (type, arg, userId, doc) {
        var deny = !Roles.userIsInRole(userId, ['client']);
        return deny;
      }
    });



    Security.defineMethod('isUser', {
      fetch: [],
      transform: null,
      deny: function (type, arg, userId, doc) {
        var deny = !Roles.userIsInRole(userId, ['user']);
        return deny;
      }
    });


    Security.defineMethod('ownsDocument', {
      fetch: ['owner'],
      transform: null,
      deny: function (type, arg, userId, doc) {
        var deny = userId !== doc.owner;
        return deny;
      }
    });

    Security.permit(['insert', 'update', 'remove'])
        .collections([UserProfiles, Stampcards, Eventcards, Audiences, AudienceStampcards, AudienceEventcards]).ifLoggedIn().isAdmin().apply();

    Security.permit(['update'])
        .collections([UserProfiles, Clients, Stampcards, Eventcards, Audiences, AudienceStampcards, AudienceEventcards]).ifLoggedIn().ownsDocument().apply();

    Security.permit(['insert'])
        .collections([Clients, Stampcards, Eventcards, Audiences, AudienceStampcards, AudienceEventcards]).ifLoggedIn().isClient().apply();

    Security.permit(['insert'])
        .collections([Audiences, AudienceStampcards, AudienceEventcards]).ifLoggedIn().isUser().apply();

    Security.permit(['insert'])
        .collections([UserProfiles]).ifLoggedIn().isUser().apply();
  }
}
```

I am not a Meteor security expert either. I wonder if it is okay to implement such middleware-wise authentication. 

I would issue a pull request about this when I am not so busy.
